# 多层感知机的简洁实现

下面我们使用Gluon来实现上一节中的多层感知机。首先导入所需的包或模块。

1.定义模型并初始化
2.读取数据并训练模型

In [1]:
import d2lzh as d2l
from mxnet import gluon, init
from mxnet.gluon import loss as gloss, nn

## 1.定义模型并初始化

和softmax回归唯一的不同在于，我们多加了一个全连接层作为隐藏层。它的隐藏单元个数为256，并使用ReLU函数作为激活函数。

In [2]:
net = nn.Sequential()
net.add(nn.Dense(256, activation='relu'),
#         nn.Dense(32, activation='relu'),
        nn.Dense(10))
net.initialize(init.Normal(sigma=0.01))

## 2.读取数据并训练模型

我们使用与[“softmax回归的简洁实现”](softmax-regression-gluon.ipynb)一节中训练softmax回归几乎相同的步骤来读取数据并训练模型。

In [3]:
batch_size = 256
train_iter, test_iter = d2l.load_data_fashion_mnist(batch_size)

loss = gloss.SoftmaxCrossEntropyLoss()
trainer = gluon.Trainer(net.collect_params(), 'sgd', {'learning_rate': 0.5})
num_epochs = 5
d2l.train_ch3(net, train_iter, test_iter, loss, num_epochs, batch_size, None,
              None, trainer)

epoch 1, loss 0.7844, train acc 0.707, test acc 0.822
epoch 2, loss 0.4906, train acc 0.818, test acc 0.843
epoch 3, loss 0.4362, train acc 0.838, test acc 0.850
epoch 4, loss 0.3947, train acc 0.853, test acc 0.858
epoch 5, loss 0.3668, train acc 0.865, test acc 0.867


## 小结

* 通过Gluon可以更简洁地实现多层感知机。

## 练习

* 尝试多加入几个隐藏层，对比上一节中从零开始的实现。
* 使用其他的激活函数，看看对结果的影响。



## 扫码直达[讨论区](https://discuss.gluon.ai/t/topic/738)

![](../img/qr_mlp-gluon.svg)

1.答：情况大致一样

2.激活函数：

'relu'：{epoch 1, loss 0.8450, train acc 0.687, test acc 0.820
epoch 2, loss 0.4873, train acc 0.819, test acc 0.851
epoch 3, loss 0.4330, train acc 0.840, test acc 0.862
epoch 4, loss 0.3979, train acc 0.853, test acc 0.863
epoch 5, loss 0.3732, train acc 0.863, test acc 0.869}


'sigmoid'：{epoch 1, loss 1.0597, train acc 0.605, test acc 0.770
epoch 2, loss 0.5773, train acc 0.786, test acc 0.808
epoch 3, loss 0.5084, train acc 0.814, test acc 0.827
epoch 4, loss 0.4696, train acc 0.830, test acc 0.841
epoch 5, loss 0.4487, train acc 0.838, test acc 0.836}

'softrelu'：epoch 1, loss 0.9588, train acc 0.652, test acc 0.803
epoch 2, loss 0.5220, train acc 0.805, test acc 0.836
epoch 3, loss 0.4634, train acc 0.829, test acc 0.847
epoch 4, loss 0.4317, train acc 0.841, test acc 0.844
epoch 5, loss 0.4118, train acc 0.850, test acc 0.863

'softsign'：{epoch 1, loss 0.7416, train acc 0.730, test acc 0.823
epoch 2, loss 0.5101, train acc 0.813, test acc 0.832
epoch 3, loss 0.4600, train acc 0.831, test acc 0.843
epoch 4, loss 0.4348, train acc 0.841, test acc 0.846
epoch 5, loss 0.4155, train acc 0.848, test acc 0.858} 

'tanh'{epoch 1, loss 0.7739, train acc 0.713, test acc 0.809
epoch 2, loss 0.5221, train acc 0.809, test acc 0.831
epoch 3, loss 0.4636, train acc 0.832, test acc 0.856
epoch 4, loss 0.4304, train acc 0.843, test acc 0.849
epoch 5, loss 0.4134, train acc 0.849, test acc 0.855}